File: android/src/main/java/com/arcticfox/txplayer/FTXPlayer.java
package com.arcticfox.txplayer;

import android.graphics.SurfaceTexture;
import android.os.Bundle;
import android.util.Log;
import android.view.Surface;

import io.flutter.plugin.common.EventChannel;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.PluginRegistry;
import io.flutter.view.TextureRegistry;

import com.tencent.rtmp.ITXLivePlayListener;
import com.tencent.rtmp.ITXVodPlayListener;
import com.tencent.rtmp.TXLiveConstants;
import com.tencent.rtmp.TXLivePlayConfig;
import com.tencent.rtmp.TXLivePlayer;
import com.tencent.rtmp.TXVodPlayer;

import java.util.HashMap;
import java.util.Map;
import java.util.Set;
import java.util.concurrent.atomic.AtomicInteger;

public class FTXPlayer implements MethodChannel.MethodCallHandler, ITXLivePlayListener, ITXVodPlayListener {
    private static final AtomicInteger mAtomicId = new AtomicInteger(0);

    final private PluginRegistry.Registrar registrar;
    private final int mPlayerId;
    private final EventChannel mEventChannel;
    private TextureRegistry.SurfaceTextureEntry mSurfaceTextureEntry;
    final private MethodChannel mMethodChannel;
    private SurfaceTexture mSurfaceTexture;
    private Surface mSurface;

    final private QueuingEventSink mEventSink = new QueuingEventSink();
    private TXVodPlayer mVodPlayer;
    private TXLivePlayConfig mLivePlayConfig;


    public FTXPlayer(PluginRegistry.Registrar registrar) {
        this.registrar = registrar;
        mPlayerId = mAtomicId.incrementAndGet();
        mMethodChannel = new MethodChannel(registrar.messenger(), FlutterTXPlayerPlugin.channelName + "/" + mPlayerId);
        mMethodChannel.setMethodCallHandler(this);

        mEventChannel = new EventChannel(registrar.messenger(), FlutterTXPlayerPlugin.channelName + "/event/" + mPlayerId);
        mEventChannel.setStreamHandler(new EventChannel.StreamHandler() {
            @Override
            public void onListen(Object o, EventChannel.EventSink eventSink) {
                mEventSink.setEventSinkProxy(eventSink);
            }

            @Override
            public void onCancel(Object o) {
                mEventSink.setEventSinkProxy(null);
            }
        });
    }

    private TXLivePlayer mLivePlayer;

    private static final int Uninitialized = -101;

    @Override
    public void onMethodCall(MethodCall call, MethodChannel.Result result) {
        if (call.method.equals("init")) {
            boolean onlyAudio = call.argument("onlyAudio");
            int playerType = call.argument("playerType");
            long id = init(onlyAudio, playerType);
            result.success(id);
        } else if (call.method.equals("play")) {
            String url = call.argument("url");
            Integer type = call.argument("playType");
            int r = startPlay(url, type == null ? -1 : type);
            result.success(r);
        } else if (call.method.equals("stop")) {
            Boolean isNeedClear = call.argument("isNeedClear");
            int r = stopPlay(isNeedClear);
            result.success(r);
        } else if (call.method.equals("isPlaying")) {
            boolean r = isPlaying();
            result.success(r);
        } else if (call.method.equals("pause")) {
            pause();
            result.success(null);
        } else if (call.method.equals("resume")) {
            resume();
            result.success(null);
        } else if (call.method.equals("setMute")) {
            boolean mute = call.argument("mute");
            setMute(mute);
            result.success(null);
        } else if (call.method.equals("setVolume")) {
            Integer volume = call.argument("volume");
            setVolume(volume);
            result.success(null);
        } else if (call.method.equals("setLiveMode")) {
            Integer type = call.argument("type");
            setLiveMode(type);
            result.success(null);
        } else if (call.method.equals("setLoop")) {
            boolean loop = call.argument("loop");
            setLoop(loop);
            result.success(null);
        }else {
            result.notImplemented();
        }
    }

    protected long init(boolean onlyAudio, int playerType) {
        if (playerType == 0) {
            if (mLivePlayer == null) {
                mLivePlayer = new TXLivePlayer(registrar.context());
                mLivePlayer.setPlayListener(this);
                setPlayer(onlyAudio);
            }
        } else {
            if (mVodPlayer == null) {
                mVodPlayer = new TXVodPlayer(registrar.context());
                mVodPlayer.setVodListener(this);
                setPlayer(onlyAudio);
            }
        }
//        Log.d("AndroidLog", "textureId :" + mSurfaceTextureEntry.id());
        return mSurfaceTextureEntry == null ? -1 : mSurfaceTextureEntry.id();
    }

    void setPlayer(boolean onlyAudio) {
        if (!onlyAudio) {
            TextureRegistry textureRegistry = registrar.textures();
            mSurfaceTextureEntry = textureRegistry.createSurfaceTexture();
            mSurfaceTexture = mSurfaceTextureEntry.surfaceTexture();
            mSurface = new Surface(mSurfaceTexture);

            if (mLivePlayer != null) {
                mLivePlayer.setSurface(mSurface);
                mLivePlayer.setSurfaceSize(1024,576);
                mLivePlayer.enableHardwareDecode(true);
            }

            if (mVodPlayer != null) {
                mVodPlayer.setSurface(mSurface);
                mVodPlayer.enableHardwareDecode(true);
            }
        }
    }

    protected void release() {
        if (mLivePlayer != null) {
            mLivePlayer.stopPlay(true);
            mLivePlayer = null;
        }

        if (mVodPlayer != null) {
            mVodPlayer.stopPlay(true);
            mVodPlayer = null;
        }

        if (mSurfaceTextureEntry != null) {
            mSurfaceTextureEntry.release();
            mSurfaceTextureEntry = null;
        }
        if (mSurfaceTexture != null) {
            mSurfaceTexture.release();
            mSurfaceTexture = null;
        }
        if (mSurface != null) {
            mSurface.release();
            mSurface = null;
        }
        mMethodChannel.setMethodCallHandler(null);
        mEventChannel.setStreamHandler(null);
    }

    void setLiveMode(int type) {
        if (mLivePlayConfig ==null) mLivePlayConfig = new TXLivePlayConfig();
        if (type == 0) {
            //自动模式
            mLivePlayConfig.setAutoAdjustCacheTime(true);
            mLivePlayConfig.setMinAutoAdjustCacheTime(1);
            mLivePlayConfig.setMaxAutoAdjustCacheTime(5);
        } else if (type == 1) {
            //极速模式
            mLivePlayConfig.setAutoAdjustCacheTime(true);
            mLivePlayConfig.setMinAutoAdjustCacheTime(1);
            mLivePlayConfig.setMaxAutoAdjustCacheTime(1);
        } else {
            //流畅模式
            mLivePlayConfig.setAutoAdjustCacheTime(false);
            mLivePlayConfig.setMinAutoAdjustCacheTime(5);
            mLivePlayConfig.setMaxAutoAdjustCacheTime(5);
        }

        if (mLivePlayer != null){
            mLivePlayer.setConfig(mLivePlayConfig);
        }
    }


    int startPlay(String url, int playType) {
        if (mLivePlayer != null) {
            return mLivePlayer.startPlay(url, playType);
        } else if (mVodPlayer != null) {
            return mVodPlayer.startPlay(url);
        }
        return Uninitialized;
    }

    int stopPlay(boolean isNeedClearLastImg) {
        if (mLivePlayer != null) {
            return mLivePlayer.stopPlay(isNeedClearLastImg);
        } else if (mVodPlayer != null) {
            return mVodPlayer.stopPlay(isNeedClearLastImg);
        }
        return Uninitialized;
    }

    boolean isPlaying() {
        if (mLivePlayer != null) {
            return mLivePlayer.isPlaying();
        } else if (mVodPlayer != null) {
            return mVodPlayer.isPlaying();
        }
        return false;
    }

    void pause() {
        if (mLivePlayer != null) {
            mLivePlayer.pause();
        } else if (mVodPlayer != null) {
            mVodPlayer.pause();
        }
    }

    void resume() {
        if (mLivePlayer != null) {
            mLivePlayer.resume();
        } else if (mVodPlayer != null) {
            mVodPlayer.resume();
        }
    }

    void setMute(boolean mute) {
        if (mLivePlayer != null) {
            mLivePlayer.setMute(mute);
        } else if (mVodPlayer != null) {
            mVodPlayer.setMute(mute);
        }
    }

    void setVolume(int volume) {
        if (mLivePlayer != null) {
            mLivePlayer.setVolume(volume);
        } else if (mVodPlayer != null) {
//            mVodPlayer.setVolume(volume);
        }
    }

    void setLoop(boolean loop){
        if (mVodPlayer != null){
            mVodPlayer.setLoop(loop);
        }
    }

    void setConfig() {
        mLivePlayConfig = new TXLivePlayConfig();
    }

    public int getPlayerId() {
        return mPlayerId;
    }


    /* see: https://cloud.tencent.com/document/product/454/7886#.E6.92.AD.E6.94.BE.E4.BA.8B.E4.BB.B6 */
    @Override
    public void onPlayEvent(int event, Bundle bundle) {
        switch (event) {
            case TXLiveConstants.PLAY_EVT_CONNECT_SUCC: //已经连接服务器
            case TXLiveConstants.PLAY_EVT_RTMP_STREAM_BEGIN: //已经连接服务器，开始拉流（仅播放 RTMP 地址时会抛送）
            case TXLiveConstants.PLAY_EVT_RCV_FIRST_I_FRAME: //收到首帧数据，越快收到此消息说明链路质量越好
            case TXLiveConstants.PLAY_EVT_PLAY_BEGIN: //视频播放开始，如果您自己做 loading，会需要它
            case TXLiveConstants.PLAY_EVT_PLAY_END: //播放结束，HTTP-FLV 的直播流不抛这个事件
            case TXLiveConstants.PLAY_ERR_NET_DISCONNECT: //网络断连，且经多次重连亦不能恢复，更多重试请自行重启播放
            case TXLiveConstants.PLAY_EVT_CHANGE_RESOLUTION: //视频分辨率发生变化（分辨率在 EVT_PARAM 参数中）

            case TXLiveConstants.PLAY_WARNING_RECONNECT:
            case TXLiveConstants.PLAY_WARNING_DNS_FAIL:
            case TXLiveConstants.PLAY_WARNING_SEVER_CONN_FAIL:
            case TXLiveConstants.PLAY_WARNING_SHAKE_FAIL:
                mEventSink.success(getParams(event, bundle));
                break;
            default:
        }
    }

    @Override
    public void onPlayEvent(TXVodPlayer txVodPlayer, int i, Bundle bundle) {

    }


    private Map<String, Object> getParams(int event, Bundle bundle) {
        Map<String, Object> param = new HashMap();
        param.put("event", event);

        if (bundle != null && !bundle.isEmpty()) {
            Set<String> keySet = bundle.keySet();
            for (String key : keySet) {
                Object val = bundle.get(key);
                param.put(key, val);
            }
        }

        return param;
    }

    @Override
    public void onNetStatus(Bundle bundle) {

    }

    @Override
    public void onNetStatus(TXVodPlayer txVodPlayer, Bundle bundle) {

    }
}

File: ios/Classes/FTXPlayer.m
//
//  FTXPlayer.m
//  Runner
//
//  Created by arcticfox on 2020/1/3.
//  Copyright © 2020 The Chromium Authors. All rights reserved.
//

#import <Foundation/Foundation.h>
#import <libkern/OSAtomic.h>
#import <TXLiteAVSDK_Player/TXLiteAVSDK.h>
#import <stdatomic.h>
#import "FTXPlayer.h"
#import "QueuingEventSink.h"

static atomic_int atomicId = 0;
static const int uninitialized = -1;

@implementation FTXPlayer{
    QueuingEventSink *_eventSink;
    FlutterMethodChannel *_methodChannel;
    FlutterEventChannel *_eventChannel;

    id<FlutterPluginRegistrar> _registrar;
//    id<FlutterTextureRegistry> _textureRegistry;
//
//    CVPixelBufferRef volatile _latestPixelBuffer;
//    CVPixelBufferRef _lastBuffer;
    
    TXLivePlayer *_txLivePlayer;
    TXVodPlayer *_txVodPlayer;
}

- (instancetype)initWithRegistrar:(id<FlutterPluginRegistrar>)registrar {
    self = [super init];
    if (self) {
        _registrar = registrar;
        int pid = atomic_fetch_add(&atomicId, 1);
        _playerId = @(pid);
        _eventSink = [QueuingEventSink new];
//        _latestPixelBuffer = nil;
//        _lastBuffer = nil;


        _methodChannel = [FlutterMethodChannel
            methodChannelWithName:[@"arcticfox.com/txplayer/"
                                      stringByAppendingString:[_playerId
                                                                  stringValue]]
                  binaryMessenger:[registrar messenger]];

        __block typeof(self) weakSelf = self;
        [_methodChannel setMethodCallHandler:^(FlutterMethodCall *call,
                                               FlutterResult result) {
          [weakSelf handleMethodCall:call result:result];
        }];

        _eventChannel = [FlutterEventChannel
            eventChannelWithName:[@"arcticfox.com/txplayer/event/"
                                     stringByAppendingString:[_playerId
                                                                 stringValue]]
                 binaryMessenger:[registrar messenger]];

        [_eventChannel setStreamHandler:self];
    }

    return self;
}

- (void)handleMethodCall:(FlutterMethodCall*)call result:(FlutterResult)result {
    NSDictionary *args = call.arguments;
    
    if([@"init" isEqualToString:call.method]){
        int playerType = [args[@"playerType"] intValue];
        BOOL onlyAudio = [args[@"onlyAudio"] boolValue];
        NSNumber* textureId = [self createPlayer:playerType onlyAudio:onlyAudio];
        result(textureId);
    }else if([@"play" isEqualToString:call.method]){
        NSString *url = args[@"url"];
        int type = -1;
        if(![[args objectForKey:@"playType"] isEqual:[NSNull null]]){
            type = [args[@"playType"] intValue];
        }
        int r = [self startPlay:url type:type];
        result(@(r));
    }else if([@"stop" isEqualToString:call.method]){
        BOOL r = [self stopPlay];
        result([NSNumber numberWithBool:r]);
    }else if([@"isPlaying" isEqualToString:call.method]){
        result([NSNumber numberWithBool:[self isPlaying]]);
    }else if([@"pause" isEqualToString:call.method]){
        [self pause];
        result(nil);
    }else if([@"resume" isEqualToString:call.method]){
        [self resume];
        result(nil);
    }else if([@"setMute" isEqualToString:call.method]){
        BOOL mute = [args[@"mute"] boolValue];
        [self setMute:mute];
        result(nil);
    }else if([@"setVolume" isEqualToString:call.method]){
        int volume = [args[@"volume"] intValue];
        [self setVolume:volume];
        result(nil);
    }else if([@"setLiveMode" isEqualToString:call.method]){
        int type = [args[@"type"] intValue];
        [self setLiveMode:type];
        result(nil);
    }else if([@"setLoop" isEqualToString:call.method]){
        BOOL loop = [args[@"loop"] boolValue];
        [self setLoop:loop];
        result(nil);
    }else {
      result(FlutterMethodNotImplemented);
    }
}


- (FlutterError *_Nullable)onListenWithArguments:(id _Nullable)arguments
                                       eventSink:
                                           (nonnull FlutterEventSink)events {
    [_eventSink setDelegate:events];
    return nil;
}

- (FlutterError *_Nullable)onCancelWithArguments:(id _Nullable)arguments {
    [_eventSink setDelegate:nil];
    return nil;
}

- (NSNumber*)createPlayer:(int)playerType onlyAudio:(BOOL)onlyAudio{
    if (playerType == 0) {
        if (_txLivePlayer == nil) {
            _txLivePlayer = [TXLivePlayer new];
            _txLivePlayer.delegate = self;
        }
    }else{
        if (_txVodPlayer == nil){
            _txVodPlayer = [TXVodPlayer new];
            _txVodPlayer.vodDelegate = self;
        }
    }
    return [NSNumber numberWithInt:-1];
}

- (int)startPlay:(NSString *)url type:(TX_Enum_PlayType)playType  {
    if (_txLivePlayer != nil) {
        return [_txLivePlayer startPlay:url type:playType];
    }else if(_txVodPlayer != nil){
        return [_txVodPlayer startPlay:url];
    }
    return uninitialized;
}

- (BOOL)stopPlay{
    if (_txLivePlayer != nil) {
        return [_txLivePlayer stopPlay];
    }else if(_txVodPlayer != nil){
        return [_txVodPlayer stopPlay];
    }
    return NO;
}

-(BOOL)isPlaying{
    if (_txLivePlayer != nil) {
        return [_txLivePlayer isPlaying];
    }else if(_txVodPlayer != nil){
        return [_txVodPlayer isPlaying];
    }
    return NO;
}

-(void)pause{
    if (_txLivePlayer != nil) {
        return [_txLivePlayer pause];
    }else if(_txVodPlayer != nil){
        return [_txVodPlayer pause];
    }
}

-(void)resume{
    if (_txLivePlayer != nil) {
        return [_txLivePlayer resume];
    }else if(_txVodPlayer != nil){
        return [_txVodPlayer resume];
    }
}

- (void)setMute:(BOOL)bEnable {
    if (_txLivePlayer != nil) {
        return [_txLivePlayer setMute:bEnable];
    }else if(_txVodPlayer != nil){
        return [_txVodPlayer setMute:bEnable];
    }
}

- (void)setVolume:(int)volume{
    if (_txLivePlayer != nil) {
        return [_txLivePlayer setVolume:volume];
    }
}

-(void)setLiveMode:(int)type{
    TXLivePlayConfig*  _config = [TXLivePlayConfig new];
    
    if (type == 0) {
        //自动模式
        _config.bAutoAdjustCacheTime   = YES;
        _config.minAutoAdjustCacheTime = 1;
        _config.maxAutoAdjustCacheTime = 5;
    }else if(type == 1){
        //极速模式
        _config.bAutoAdjustCacheTime   = YES;
        _config.minAutoAdjustCacheTime = 1;
        _config.maxAutoAdjustCacheTime = 1;
    }else{
        //流畅模式
        _config.bAutoAdjustCacheTime   = NO;
        _config.minAutoAdjustCacheTime = 5;
        _config.maxAutoAdjustCacheTime = 5;
    }
    [_txLivePlayer setConfig:_config];
}

-(void)setLoop:(BOOL)bLoop{
    if (_txVodPlayer != nil) {
        _txVodPlayer.loop = bLoop;
    }
}

/**
 * 直播事件通知
 * @param EvtID 参见 TXLiveSDKEventDef.h
 * @param param 参见 TXLiveSDKTypeDef.h
 */
- (void)onPlayEvent:(int)EvtID withParam:(NSDictionary *)param{
    switch (EvtID) {
        case PLAY_EVT_CONNECT_SUCC: //已经连接服务器
        case PLAY_EVT_RTMP_STREAM_BEGIN: //已经连接服务器，开始拉流（仅播放 RTMP 地址时会抛送）
        case PLAY_EVT_RCV_FIRST_I_FRAME: //收到首帧数据，越快收到此消息说明链路质量越好
        case PLAY_EVT_PLAY_BEGIN: //视频播放开始，如果您自己做 loading，会需要它
        case PLAY_EVT_PLAY_END: //播放结束，HTTP-FLV 的直播流不抛这个事件
        case PLAY_ERR_NET_DISCONNECT: //网络断连，且经多次重连亦不能恢复，更多重试请自行重启播放
        case PLAY_EVT_CHANGE_RESOLUTION: //视频分辨率发生变化（分辨率在 EVT_PARAM 参数中）
        case PLAY_WARNING_RECONNECT:
        case PLAY_WARNING_DNS_FAIL:
        case PLAY_WARNING_SEVER_CONN_FAIL:
        case PLAY_WARNING_SHAKE_FAIL:
            [_eventSink success:[FTXPlayer getParamsWithEvent:EvtID withParams:param]];
            break;
        default:
            break;
    }
}

+(NSDictionary *)getParamsWithEvent:(int)EvtID withParams:(NSDictionary *)params{
    NSMutableDictionary<NSString*,NSObject*> *dict = [NSMutableDictionary dictionaryWithObject:@(EvtID) forKey:@"event"];
    
    if (params != nil && params.count != 0) {
        for (NSString *key in params) {
            NSObject* value = params[key];
            dict[key] = value;
        }
    }
    return dict;
}
/**
 * 网络状态通知
 * @param param 参见 TXLiveSDKTypeDef.h
 */
- (void)onNetStatus:(NSDictionary *)param{
    
}

/**
 * 点播事件通知
 *
 * @param player 点播对象
 * @param EvtID 参见TXLiveSDKTypeDef.h
 * @param param 参见TXLiveSDKTypeDef.h
 * @see TXVodPlayer
 */
-(void) onPlayEvent:(TXVodPlayer *)player event:(int)EvtID withParam:(NSDictionary*)param{
    
}

/**
 * 网络状态通知
 *
 * @param player 点播对象
 * @param param 参见TXLiveSDKTypeDef.h
 * @see TXVodPlayer
 */
-(void) onNetStatus:(TXVodPlayer *)player withParam:(NSDictionary*)param{
    
}

-(void)destroy{
    [self stopPlay];
    _txLivePlayer = nil;
    _txVodPlayer = nil;
    
    [_methodChannel setMethodCallHandler:nil];
    _methodChannel = nil;

    [_eventSink setDelegate:nil];
    _eventSink = nil;
    [_eventChannel setStreamHandler:nil];
    _eventChannel = nil;
}
@end
